Integrate eslint with eslint-config-standard

kaisalmen opened the pull request:
I have based this branch on your branch, because otherwise the changes are obsolete again soon and would need to be redone. Therefore the target branch is currently also `use-monaco-vscode-api`

@CGNonofr I used `eslint-config-standard`, but I made some adjustments:
- Indent is 4 instead of 2
- I favor semi-colons
- `no-dupe-class-member` is off
- `no-redeclare`, `no-useless-constructor` and `no-void` have been reduced to **warn** (you should check if the code should or could be changed or if `eslint-disable` should be set instead and we can then remove the rule reduction)

This closes #366

CGNonofr commented on .eslintrc.js:
```diff
@@ -0,0 +1,37 @@
+        semi: ['error', 'always'],
+        'no-extra-semi': 'off',
+        '@typescript-eslint/no-extra-semi': ['error'],
+        'no-dupe-class-members': 'off',
```
That's the base javascript rule that doesn't work well with typescript, what about enabling `@typescript-eslint/no-dupe-class-members`?

CGNonofr commented on .eslintrc.js:
```diff
@@ -0,0 +1,37 @@
+        'no-extra-semi': 'off',
+        '@typescript-eslint/no-extra-semi': ['error'],
+        'no-dupe-class-members': 'off',
+        'no-redeclare': 'warn',
```
Same, what about enabling `@typescript-eslint/no-redeclare`?

kaisalmen replied:
Same as above. Thanks 🙂

CGNonofr commented on .github/workflows/actions.yml:
```diff
@@ -17,4 +17,6 @@ jobs:
       shell: bash
       run: |
-        npm i && npm run build
+        npm i
```
Why is it `npm i` instead of `npm ci`?

kaisalmen replied:
Good point

CGNonofr commented on packages/client/src/monaco-language-client.ts:
```diff
@@ -52,28 +54,28 @@ export class MonacoLanguageClient extends BaseLanguageClient {
+            });
+        };
+        const originalHandleUnregistrationRequest: (params: UnregistrationParams) => Promise<void> = this['handleRegistrationRequest'].bind(this);
+        this['handleRegistrationRequest'] = (params: UnregistrationParams) => {
```
I don't think it should be changed

kaisalmen replied:
Sorry, stupid copy paste issue, when I changed the code back (manually)

CGNonofr commented:
Btw, you can use the `header/header` rule to ensure all files have the proper header

kaisalmen commented:
I implemented the review comments and rebased the branch. Now, the build will fail because `vscode-ws-jsonrpc` will have lint related errors. I will fix this afternoon, I think.

kaisalmen commented:
I have now enabled eslint globally for all packages and thanks for mentioning `header/header`. I have now integrated it.

kaisalmen commented:
Build is green. Everything is in. If you agree with the latest changes this is good to go, I think.

CGNonofr commented on .github/workflows/actions.yml:
```diff
@@ -17,4 +17,6 @@ jobs:
-        npm i && npm run build
+        npm ci
+        npm run build
+        npm run lint
```
Isn't it different? what happens if the npm ci fails?

CGNonofr replied:
Nice to know :+1:

CGNonofr commented on packages/client/src/disposable.ts:
```diff
@@ -3,31 +3,29 @@
+    dispose (): void {
         while (this.disposables.length !== 0) {
-            this.disposables.pop()!.dispose();
+            this.disposables.pop()?.dispose();
```
What was the issue with the `!`?

kaisalmen replied:
None, my error, happend after first round of lint issue resolution. I will change back.

CGNonofr commented on packages/client/src/monaco-language-client.ts:
```diff
@@ -2,81 +2,83 @@
  * Licensed under the MIT License. See License.txt in the project root for license information.
  * ------------------------------------------------------------------------------------------ */
+
+/* eslint-disable dot-notation */
```
Are you sure it's a good idea to disable a rule for a whole file?

kaisalmen replied:
No, it should not be there. I changed the config to this now (upcoming commit)
```javascript
        'dot-notation': 'off',
        '@typescript-eslint/dot-notation': ['error']
```
but now VSCode is auto-formatting the code to dotted notation all the time and this leads to TypeScript errors, for example:

```shell
src/monaco-language-client.ts:55:103 - error TS2341: Property 'handleRegistrationRequest' is private and only accessible within class 'BaseLanguageClient'.

55         const originalHandleRegistrationRequest: (params: RegistrationParams) => Promise<void> = this.handleRegistrationRequest.bind(this);
```

CGNonofr replied:
Yes, this is a known hack to access a private field in typescript!

The best way is to disable the dot-notation error on each line I think

kaisalmen replied:
Did this now. See new commit

CGNonofr commented on packages/examples/browser-lsp/src/serverWorker.ts:
```diff
@@ -67,22 +66,21 @@ function getColorInformation(textDocument: TextDocumentIdentifier) {
+function getColorPresentation (color: Color, range: Range) {
     const result: ColorPresentation[] = [];
-    const red256 = Math.round(color.red * 255), green256 = Math.round(color.green * 255), blue256 = Math.round(color.blue * 255);
+    const red256 = Math.round(color.red * 255); const green256 = Math.round(color.green * 255); const blue256 = Math.round(color.blue * 255);
```
I'm surprised there is no eslint rule preventing putting multiple variables on the same line

kaisalmen replied:
No, it should not be there. I changed the config to this now (upcoming commit)
```javascript
        'dot-notation': 'off',
        '@typescript-eslint/dot-notation': ['error']
```
but now VSCode is auto-formatting the code to dotted notation all the time and this leads to TypeScript errors, for example:

```shell
src/monaco-language-client.ts:55:103 - error TS2341: Property 'handleRegistrationRequest' is private and only accessible within class 'BaseLanguageClient'.

55         const originalHandleRegistrationRequest: (params: RegistrationParams) => Promise<void> = this.handleRegistrationRequest.bind(this);
```

kaisalmen replied:
At least it is not configured in `standard`. 🙂

CGNonofr commented on packages/vscode-ws-jsonrpc/src/disposable.ts:
```diff
@@ -1,33 +1,33 @@
-
-}
\ No newline at end of file
+}
```
(not related to this PR) Are you sure it's relevant to have the same code duplicated in the client and vscode-ws-jsonrpc?

kaisalmen replied:
Not nice, but both packages are independent. Only the examples use both.

kaisalmen replied:
I actually have a good idea for that.

kaisalmen replied:
I forced another update. At least both `disposable.ts` are identical now.

CGNonofr commented:
> I will produce new `dev` releases once this is merged. Sounds good?

Sounds good! What will we wait before releasing a final version?

CGNonofr reviewed (approved):
LGTM

kaisalmen commented:
> Sounds good! What will we wait for before releasing a final version?

Actually a good question. We don't really have official beta tester. And a major version change should not immediately cripple correctly configured projects depending on here.

CGNonofr commented:
It seems pretty safe to release a major version to me, and release fixes later, or we won't get any feedback

I'm already testing the new vscode-api approch for more than a week, and everything else is just refactors, it can't break things, can it?

kaisalmen commented:
> it can't break things, can it?

Let's go

kaisalmen commented:
Done 🎉 

https://www.npmjs.com/package/monaco-languageclient/v/2.0.0

https://www.npmjs.com/package/vscode-ws-jsonrpc/v/1.0.0

kaisalmen commented:
We should archive the repos of `@codingame/monaco-jsonrpc` and `vscode-ws-jsonrpc`  and add links in the README that link to here.

CGNonofr commented:
done!

kaisalmen commented:
✅
